### DISCUSSION TITLE
How to call a command from outside its class in VS Community Toolkit

### DISCUSSION THREAD

**Discussion by ghost** [VsixCommunity/Community.VisualStudio.Toolkit#520] [Q&A] [answered]
I have a command named Login: 

[Command(PackageIds.Login)]
internal sealed class Login : BaseCommand<Login>
{
    protected override async Task ExecuteAsync(OleMenuCmdEventArgs e)
    {
    
Inside this ExecuteAsync some logic is performed so that if a singleton class contains or not information about the login, the button shows "Login" or "Logout" and the procedure is of course, different. 
How can I access this method from other parts of my program? 

I already have a hint, since inside the BaseCommand interface there is a method called InitializeAsync that says: Initializes the command. This method must be called from the <see cref="AsyncPackage.InitializeAsync"/> method for the command to work.

**Comment by RobertvanderHulst** [comment, marked as answer]
When you call RegisterCommandsAsync() at the package level, this method uses reflection to find all subclasses of BaseCommand and create and initialize the commands by calling InitializeAsync(). You can add any code that you want in this method, including code to store a reference to the singleton. Later in BeforeQueryStatus() and Execute() you are free to do whatever you want.

**Reply by ghost** [reply]
Hi Robert and thank you for the quick response. 

Your response lead me to finding a solution for my problem, so I will write it here in case anyone finds it useful. 

I was able to override InitializeCompletedAsync() inside my Login command class. There, I populate a new field; 

        public static Login Instance { get; private set; }
        protected override async Task InitializeCompletedAsync()
        {
            Instance = this;
        }
Now I can create a public method performing any operation needed, changing a command.text or doing whatever is needed. This method can be called from every part of the project this way; Login.Instance.Method();
